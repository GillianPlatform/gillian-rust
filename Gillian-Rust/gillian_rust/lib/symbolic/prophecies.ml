open Monadic
module DR = Delayed_result
open DR.Syntax
open Delayed.Syntax
open Heap
open Gil_syntax
open Delayed_utils

type prophecy = {
  value : Expr.t;
  obs_ctl : Expr.t;
  observer : bool;
  controller : bool;
}

let prophecy_lvars prophecy =
  let open Utils.Containers in
  SS.union (Expr.lvars prophecy.obs_ctl) (Expr.lvars prophecy.value)

let pp_prophecy =
  let open Fmt in
  Fmt.braces
  @@ record ~sep:semi
       [
         field "value" (fun x -> x.value) Expr.pp;
         field "obs_ctl" (fun x -> x.obs_ctl) Expr.pp;
         field "observer" (fun x -> x.observer) Fmt.bool;
         field "controller" (fun x -> x.controller) Fmt.bool;
       ]

type t = prophecy MemMap.t

let sure_is_nonempty =
  MemMap.exists (fun _ { observer; controller; _ } -> observer || controller)

let assertions (pcies : t) =
  let cps loc pcy =
    (* Location and ty are uniform *)
    let loc = Expr.loc_from_loc_name loc in
    (* The future value of the prophecy variable *)
    let value =
      let cp = Actions.cp_to_name Pcy_value in
      Asrt.GA (cp, [ loc ], [ pcy.value ])
    in
    let controller =
      match pcy.controller with
      | false -> []
      | true ->
          let cp = Actions.cp_to_name Pcy_controller in
          [ Asrt.GA (cp, [ loc ], [ pcy.obs_ctl ]) ]
    in
    let observer =
      match pcy.observer with
      | false -> []
      | true ->
          let cp = Actions.cp_to_name Value_observer in
          [ Asrt.GA (cp, [ loc ], [ pcy.obs_ctl ]) ]
    in
    value :: (controller @ observer)
  in
  MemMap.fold (fun loc pcy acc -> cps loc pcy @ acc) pcies []

let lvars t =
  let open Utils.Containers in
  MemMap.fold (fun _ pcy acc -> SS.union acc (prophecy_lvars pcy)) t SS.empty

let empty = MemMap.empty
let to_yojson _ = `Null
let of_yojson _ = Error "Heap.of_yojson: Not implemented"

let cons_value_obs pcy_env pcy_var =
  match MemMap.find_opt pcy_var pcy_env with
  | Some ({ observer = true; obs_ctl; _ } as pcy) ->
      (* We consume the pcy *)
      let new_pcy_env =
        MemMap.add pcy_var { pcy with observer = false } pcy_env
      in
      DR.ok (obs_ctl, new_pcy_env)
  | Some { observer = false; _ } -> DR.error (Err.Missing_observer pcy_var)
  | None -> DR.error (Err.Missing_pcy pcy_var)

let prod_value_obs pcy_env pcy_id obs_value =
  let+ new_block =
    match MemMap.find_opt pcy_id pcy_env with
    | Some { observer = true; _ } -> Delayed.vanish () (* Duplicated resource *)
    | Some { value; obs_ctl; observer = false; controller = true } ->
        (* We already have a controller, we learn equality *)
        let learned = [ Formula.Infix.( #== ) obs_value obs_ctl ] in
        Delayed.return ~learned
          { value; observer = true; controller = true; obs_ctl }
    | Some { value; observer = false; controller = false; obs_ctl = _ } ->
        (* We have neither the controller nor the observer, the obs_ctl value is irrelevant *)
        Delayed.return
          { value; observer = true; controller = false; obs_ctl = obs_value }
    | None ->
        Delayed.return
          {
            value = Expr.LVar (LVar.alloc ());
            obs_ctl = obs_value;
            observer = true;
            controller = false;
          }
  in
  MemMap.add pcy_id new_block pcy_env

let cons_controller pcy_env pcy_var =
  match MemMap.find_opt pcy_var pcy_env with
  | Some ({ controller = true; obs_ctl; _ } as pcy) ->
      (* We consume the pcy *)
      let new_pcy_env =
        MemMap.add pcy_var { pcy with controller = false } pcy_env
      in
      DR.ok (obs_ctl, new_pcy_env)
  | Some { controller = false; _ } -> DR.error (Err.Missing_controller pcy_var)
  | None -> DR.error (Err.Missing_pcy pcy_var)

let prod_controller pcy_env pcy_id ctl_value =
  let+ new_block =
    match MemMap.find_opt pcy_id pcy_env with
    | Some { controller = true; _ } ->
        Delayed.vanish () (* Duplicated resource *)
    | Some { value; obs_ctl; observer = true; controller = false } ->
        (* We already have an observer, we learn equality *)
        let learned = [ Formula.Infix.( #== ) ctl_value obs_ctl ] in
        Delayed.return ~learned
          { value; observer = true; controller = true; obs_ctl }
    | Some { value; observer = false; controller = false; obs_ctl = _ } ->
        (* We have neither the controller nor the observer, the obs_ctl value is irrelevant *)
        Delayed.return
          { value; observer = false; controller = true; obs_ctl = ctl_value }
    | None ->
        Delayed.return
          {
            value = Expr.LVar (LVar.alloc ());
            obs_ctl = ctl_value;
            observer = false;
            controller = true;
          }
  in
  MemMap.add pcy_id new_block pcy_env

let cons_value pcy_env pcy_id =
  (* The value is persistent information, it doesn't get consumed *)
  match MemMap.find_opt pcy_id pcy_env with
  | Some { value; _ } -> (value, pcy_env)
  | None ->
      (* If not there, we simply create it, and now we know its value. *)
      let value = Expr.LVar (LVar.alloc ()) in
      ( value,
        MemMap.add pcy_id
          {
            value;
            observer = false;
            controller = false;
            obs_ctl = Expr.Lit Null;
            (* The obs_ctl we assign is irrelevant, since we have neither observer nor controller *)
          }
          pcy_env )

let prod_value pcy_env pcy_id new_value =
  match MemMap.find_opt pcy_id pcy_env with
  | Some pcy ->
      (* `value` is a function, so if we already know its value, we now that whatever we have is equal *)
      Delayed.return
        ~learned:[ Formula.Infix.( #== ) pcy.value new_value ]
        pcy_env
  | None ->
      (* The obs_ctl we assign is irrelevant, since we have neither observer nor controller *)
      let block =
        {
          value = new_value;
          observer = false;
          controller = false;
          obs_ctl = Expr.Lit Null;
        }
      in
      Delayed.return (MemMap.add pcy_id block pcy_env)

let assign pcy_env pcy_var assigned =
  (* In order to assign an obs_ctl value, we need *both* the controller and observer at the same time. *)
  match MemMap.find_opt pcy_var pcy_env with
  | None -> DR.error (Err.Missing_pcy pcy_var)
  | Some { controller = false; _ } -> DR.error (Err.Missing_controller pcy_var)
  | Some { observer = false; _ } -> DR.error (Err.Missing_observer pcy_var)
  | Some ({ controller = true; observer = true; _ } as pcy) ->
      let new_pcies =
        MemMap.add pcy_var { pcy with obs_ctl = assigned } pcy_env
      in
      DR.ok new_pcies

let alloc pcy_env =
  let pcy_id = ALoc.alloc () in
  let value = Expr.LVar (LVar.alloc ()) in
  let obs_ctl = Expr.LVar (LVar.alloc ()) in
  let block = { value; obs_ctl; observer = true; controller = true } in
  let updated_env = MemMap.add pcy_id block pcy_env in
  (Expr.ALoc pcy_id, updated_env)

let pp ft t =
  let open Fmt in
  pf ft "@[<v 2>Prophecies:@,%a@]"
    (iter_bindings MemMap.iter (pair ~sep:(any " -> ") string pp_prophecy))
    t

let merge old_proph new_proph =
  let open Formula.Infix in
  let value_eq = old_proph.value #== new_proph.value in
  let controller = new_proph.controller || old_proph.controller in
  let observer = new_proph.observer || old_proph.observer in
  let obs_ctl, ctl_obs_eq =
    match
      ( old_proph.controller || old_proph.observer,
        new_proph.controller || new_proph.observer )
    with
    | true, true ->
        (* Both obs_ctl are relevant *)
        let ctl_obs_eq = [ old_proph.obs_ctl #== new_proph.obs_ctl ] in
        (new_proph.obs_ctl, ctl_obs_eq)
    | true, false ->
        (* Only old obs_ctl is relevant *)
        (old_proph.obs_ctl, [])
    | false, true ->
        (* Only new obs_ctl is relevant *)
        (new_proph.obs_ctl, [])
    | false, false ->
        (* Both obs_ctl are irrelevant *)
        (Expr.Lit Null, [])
  in
  let learned = value_eq :: ctl_obs_eq in
  DR.ok ~learned { value = new_proph.value; obs_ctl; observer; controller }

let substitution pcy_env subst =
  let open Gillian.Symbolic in
  if Subst.is_empty subst then DR.ok pcy_env
  else
    let loc_subst =
      Subst.fold subst
        (fun l r acc ->
          match l with
          | ALoc loc | Lit (Loc loc) -> (loc, r) :: acc
          | _ -> acc)
        []
    in
    let subst_expr = Subst.subst_in_expr subst ~partial:true in
    let values_substed =
      MemMap.map
        (fun block ->
          let value = subst_expr block.value in
          let obs_ctl = subst_expr block.obs_ctl in
          let controller = block.controller in
          let observer = block.observer in
          { value; obs_ctl; controller; observer })
        pcy_env
    in
    List.fold_left
      (fun acc (old_loc, new_loc) ->
        let** acc = acc in
        let new_loc =
          match new_loc with
          | Expr.Lit (Loc loc) | ALoc loc -> loc
          | _ ->
              Fmt.failwith
                "substitution failed, for location, target isn't a location"
        in
        match (MemMap.find_opt old_loc acc, MemMap.find_opt new_loc acc) with
        | None, None | None, Some _ -> DR.ok acc
        | Some prophecy, None ->
            MemMap.remove old_loc acc |> MemMap.add new_loc prophecy |> DR.ok
        | Some old_prophecy, Some new_prophecy ->
            let++ merged = merge old_prophecy new_prophecy in
            MemMap.remove old_loc acc |> MemMap.add new_loc merged)
      (DR.ok values_substed) loc_subst
