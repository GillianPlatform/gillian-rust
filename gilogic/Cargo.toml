[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }