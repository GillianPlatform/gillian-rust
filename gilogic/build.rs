fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
