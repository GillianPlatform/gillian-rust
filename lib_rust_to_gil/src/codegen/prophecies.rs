use crate::prelude::*;

use rustc_middle::{mir::Place, ty::Ty};

use gillian::gil::Expr;

use super::context::GilCtxt;

mod action_names {
    pub(crate) const PCY_ALLOC: &str = "pcy_alloc";
}

pub enum ProphecyAction<'tcx> {
    Alloc(Ty<'tcx>),
}

impl<'tcx, 'body> GilCtxt<'tcx, 'body> {
    fn push_pcy_action(&mut self, target: String, action: ProphecyAction<'tcx>) {
        match action {
            ProphecyAction::Alloc(ty) => {
                let ty = self.encode_type(ty).into();
                self.push_cmd(Cmd::Action {
                    variable: target,
                    action_name: action_names::PCY_ALLOC.to_string(),
                    parameters: vec![ty],
                })
            }
        }
    }

    fn push_alloc_prophecy(&mut self, typ: Ty<'tcx>) -> Expr {
        let action = ProphecyAction::Alloc(typ);
        let target = self.temp_var();
        self.push_pcy_action(target.clone(), action);
        Expr::PVar(target)
    }

    pub fn push_create_prophecy_for(&mut self, place: Place<'tcx>) -> Expr {
        if !self.prophecies_enabled() {
            fatal!(self, "Prophecies are not enabled, something is wrong");
        }
        let ty = self.place_ty(place).ty;
        match self.global_env().get_repr_ty_for(ty) {
            Some(repr_ty) => self.push_alloc_prophecy(repr_ty),
            None => Expr::null(),
        }
    }
}
