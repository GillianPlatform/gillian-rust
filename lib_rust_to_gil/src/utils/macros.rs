macro_rules! fatal {
  ($e: expr, $($tts:tt)*) => {
      $e.ty_ctxt.sess.fatal(&format!($($tts)*))
  };
}

pub(crate) use fatal;
