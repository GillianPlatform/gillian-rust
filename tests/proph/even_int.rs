//@check-pass
extern crate creusillian;
extern crate gilogic;

use gilogic::macros::{assertion, predicate};
use gilogic::mutref_auto_resolve;
use gilogic::prophecies::{Ownable, Prophecised};

struct EvenInt {
    num: i32,
}

impl Ownable for EvenInt {
    type RepresentationTy = i32;

    #[predicate]
    fn own(self, model: i32) {
        assertion!((self == EvenInt { num: model }) * (model % 2 == 0));
    }
}

impl EvenInt {
    #[creusillian::ensures(true)]
    pub fn new_2(x: i32) -> Self {
        if x % 2 == 0 {
            Self { num: x }
        } else {
            if x < 1000 {
                Self { num: x + 1 }
            } else {
                Self { num: x - 1 }
            }
        }
    }

    pub unsafe fn new(x: i32) -> Self {
        Self { num: x }
    }

    #[creusillian::ensures(true)]
    pub fn new_3(x: i32) -> Option<Self> {
        if x % 2 == 0 {
            let y = unsafe { Self::new(x) };
            Some(y)
        } else {
            None
        }
    }

    unsafe fn add(&mut self) {
        self.num += 1;
    }

    #[creusillian::ensures(true)]
    pub fn test(&mut self) {
        if self.num % 2 != 0 {
            panic!()
        }
    }

    #[creusillian::requires((*self@) <= i32::MAX@ - 2)]
    #[creusillian::ensures((^self@) == (*self@) + 2)]
    pub fn add_two(&mut self) {
        self.num;

        unsafe {
            self.add();
            self.add();
        }
        mutref_auto_resolve!(self);
    }
}
