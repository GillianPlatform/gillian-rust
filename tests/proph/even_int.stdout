#begin_init_data
{"EvenInt":["Struct","ReprRust",[["num",["Scalar","i32"]]]]}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_pcy.gil", "i__lang_pcy.gil";

axiomatic spec "gilogic::prophecies::prophecy_auto_update::<EvenInt>"(pLft_0, p)
[[ <value>(l-nth(l-nth(#p, 0i), 0i), l-nth(l-nth(#p, 0i), 1i), {{ "adt", "EvenInt", {{  }} }}; #v) * "<EvenInt as gilogic::prophecies::Ownable>::own"(#v, #r) * <value_observer>(l-nth(#p, 1i); #a) * <pcy_controller>(l-nth(#p, 1i); #a) * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj) * (#p == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }}) * (p == #p) * (pLft_0 == #pLft_0) ]]
[[<value>(l-nth(l-nth(#p, 0i), 0i), l-nth(l-nth(#p, 0i), 1i), {{ "adt", "EvenInt", {{  }} }}; #v) * "<EvenInt as gilogic::prophecies::Ownable>::own"(#v, #r) * <value_observer>(l-nth(#p, 1i); #r) * <pcy_controller>(l-nth(#p, 1i); #r)]]
normal

axiomatic spec "gilogic::prophecies::prophecy_resolve::<EvenInt>"(pLft_0, p)
[[ "<&mut EvenInt as gilogic::prophecies::Ownable>::own"(#pLft_0, #p, #m) * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj) * (#p == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }}) * (p == #p) * (pLft_0 == #pLft_0) ]]
[[<observation>(((l-nth(#m, 0i)) = (l-nth(#m, 1i))); )]]
normal

pred "<EvenInt as gilogic::prophecies::Ownable>::own"(+self, model):
  (#self == {{ #model }}) * (((#model) i% (2i)) == 0i) * types(#model: Int)
  * (model == #model)
  * (self == #self);


pred "<i32 as gilogic::prophecies::Ownable>::own"(+self, model):
  (-2147483648i i<=# #self) * (#self i<=# 2147483647i) * (#self == #model)
  * types(#self: Int)
  * types(#model: Int)
  * (model == #model)
  * (self == #self);


pred "<std::option::Option<EvenInt> as gilogic::prophecies::Ownable>::own"(+self, model):
  (#self == {{ 0i, {{  }} }}) * (#model == {{ 0i, {{  }} }}) * (model == #model)
  * (self == #self), (#self == {{ 1i, {{ #ax }} }})
  * "<EvenInt as gilogic::prophecies::Ownable>::own"(#ax, #repr)
  * (#model == {{ 1i, {{ #repr }} }})
  * (model == #model)
  * (self == #self);


pred "<&mut EvenInt as gilogic::prophecies::Ownable>::own"(+pLft_erased, +self, model):
  (#model == {{ #a, #b }})
  * "gilogic::prophecies::mut_ref_inner_proph::<EvenInt>"(pLft_erased, #self)
  * <value_observer>(l-nth(#self, 1i); #a)
  * (#lvar_0 == #b)
  * <pcy_value>(l-nth(#self, 1i); #lvar_0)
  * types(#lvar_1: Obj, #lvar_2: List, #lvar_3: Obj)
  * (#self == {{ {{ #lvar_1, #lvar_2 }}, #lvar_3 }})
  * (model == #model)
  * (self == #self);


pred "<() as gilogic::prophecies::Ownable>::own"(+self, model):
  (#self == {{  }}) * (#self == #model) * (model == #model) * (self == #self);


pred "gilogic::prophecies::mut_ref_inner_proph::<EvenInt>"(+pLft_a, +this):
  <value>(l-nth(l-nth(#this, 0i), 0i), l-nth(l-nth(#this, 0i), 1i), {{ "adt", "EvenInt", {{  }} }}; #v)
  * "<EvenInt as gilogic::prophecies::Ownable>::own"(#v, #repr)
  * <pcy_controller>(l-nth(#this, 1i); #repr)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#this == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (this == #this);
guard: <lft>(pLft_a; true);

spec "EvenInt::new_2"(x)
[[ "<i32 as gilogic::prophecies::Ownable>::own"(#x, #x_repr) * types(#x_repr: Int) * types(#x: Int) * (x == #x) ]]
[["<EvenInt as gilogic::prophecies::Ownable>::own"(ret, #ret_repr) * <observation>(true; )]]
normal

proc "EvenInt::new_2"(x) {
        m_1x := x;
        m_14 := [alloc]({{ "tuple", {{ "i32", "bool" }} }});
        m_17 := [alloc]({{ "tuple", {{ "i32", "bool" }} }});
  bb0:  m_4 := m_1x;
        m_5 := ((2i) = (0i));
        g_0 := m_5;
        u := "i__lang_assert"(not g_0, "remainder by zero");
        goto bb1;
  bb1:  m_6 := ((2i) = (-1i));
        m_7 := ((m_4) = (-2147483648i));
        g_1 := m_6;
        g_2 := m_7;
        m_8 := ((g_1) and (g_2));
        g_3 := m_8;
        u := "i__lang_assert"(not g_3, "overflow check failed");
        goto bb2;
  bb2:  g_4 := m_4;
        m_3 := ((g_4) i% (2i));
        g_5 := m_3;
        m_2 := ((g_5) = (0i));
        g_6 := m_2;
        goto [g_6] bb3 bb4;
  bb3:  m_9 := m_1x;
        g_7 := m_9;
        m_0 := {{ g_7 }};
        goto bb10;
  bb4:  m_11 := m_1x;
        g_8 := m_11;
        m_10 := ((g_8) i< (1000i));
        g_9 := m_10;
        goto [g_9] bb5 bb7;
  bb5:  m_13 := m_1x;
        g_10 := "i__binop_checked_add"(m_13, 1i, 2147483647i);
        u := [store_value](l-nth(m_14, 0i), l-nth(m_14, 1i), {{ "tuple", {{ "i32", "bool" }} }}, g_10);
        g_12 := [load_value](l-nth(m_14, 0i), l+(l-nth(m_14, 1i), {{ {{ "f", 1i, {{ "tuple", {{ "i32", "bool" }} }} }} }}), "bool", false);
        g_11 := l-nth(g_12, 0i);
        u := "i__lang_assert"(not g_11, "overflow check failed");
        goto bb6;
  bb6:  g_14 := [load_value](l-nth(m_14, 0i), l+(l-nth(m_14, 1i), {{ {{ "f", 0i, {{ "tuple", {{ "i32", "bool" }} }} }} }}), "i32", false);
        g_13 := l-nth(g_14, 0i);
        m_12 := g_13;
        g_15 := m_12;
        m_0 := {{ g_15 }};
        goto bb9;
  bb7:  m_16 := m_1x;
        g_16 := "i__binop_checked_sub"(m_16, 1i);
        u := [store_value](l-nth(m_17, 0i), l-nth(m_17, 1i), {{ "tuple", {{ "i32", "bool" }} }}, g_16);
        g_18 := [load_value](l-nth(m_17, 0i), l+(l-nth(m_17, 1i), {{ {{ "f", 1i, {{ "tuple", {{ "i32", "bool" }} }} }} }}), "bool", false);
        g_17 := l-nth(g_18, 0i);
        u := "i__lang_assert"(not g_17, "overflow check failed");
        goto bb8;
  bb8:  g_20 := [load_value](l-nth(m_17, 0i), l+(l-nth(m_17, 1i), {{ {{ "f", 0i, {{ "tuple", {{ "i32", "bool" }} }} }} }}), "i32", false);
        g_19 := l-nth(g_20, 0i);
        m_15 := g_19;
        g_21 := m_15;
        m_0 := {{ g_21 }};
        goto bb9;
  bb9:  goto bb10;
  bb10: g_22 := m_0;
        ret := g_22;
        goto rlab;
  rlab: u := [free](l-nth(m_14, 0i), l-nth(m_14, 1i), {{ "tuple", {{ "i32", "bool" }} }});
        u := [free](l-nth(m_17, 0i), l-nth(m_17, 1i), {{ "tuple", {{ "i32", "bool" }} }});
        return
};

proc "EvenInt::new"(x) {
        m_1x := x;
  bb0:  m_2 := m_1x;
        g_0 := m_2;
        m_0 := {{ g_0 }};
        g_1 := m_0;
        ret := g_1;
        goto rlab;
  rlab: return
};

spec "EvenInt::new_3"(x)
[[ "<i32 as gilogic::prophecies::Ownable>::own"(#x, #x_repr) * types(#x_repr: Int) * types(#x: Int) * (x == #x) ]]
[["<std::option::Option<EvenInt> as gilogic::prophecies::Ownable>::own"(ret, #ret_repr) * <observation>(true; )]]
normal

proc "EvenInt::new_3"(x) {
        m_1x := x;
  bb0:  m_4 := m_1x;
        m_5 := ((2i) = (0i));
        g_0 := m_5;
        u := "i__lang_assert"(not g_0, "remainder by zero");
        goto bb1;
  bb1:  m_6 := ((2i) = (-1i));
        m_7 := ((m_4) = (-2147483648i));
        g_1 := m_6;
        g_2 := m_7;
        m_8 := ((g_1) and (g_2));
        g_3 := m_8;
        u := "i__lang_assert"(not g_3, "overflow check failed");
        goto bb2;
  bb2:  g_4 := m_4;
        m_3 := ((g_4) i% (2i));
        g_5 := m_3;
        m_2 := ((g_5) = (0i));
        g_6 := m_2;
        goto [g_6] bb3 bb5;
  bb3:  m_10 := m_1x;
        g_7 := m_10;
        g_8 := "EvenInt::new"(g_7);
        m_9y := g_8;
        goto bb4;
  bb4:  g_9 := m_9y;
        m_11 := g_9;
        g_10 := m_11;
        m_0 := {{ 1i, {{ g_10 }} }};
        goto bb6;
  bb5:  m_0 := {{ 0i, {{  }} }};
        goto bb6;
  bb6:  g_11 := m_0;
        ret := g_11;
        goto rlab;
  rlab: return
};

proc "EvenInt::add"(pLft_0, self) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, g_0);
        m_2 := [alloc]({{ "tuple", {{ "i32", "bool" }} }});
  bb0:  g_2 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, true);
        g_1 := l-nth(g_2, 0i);
        g_4 := [load_value](l-nth(l-nth(g_1, 0i), 0i), l+(l-nth(l-nth(g_1, 0i), 1i), {{ {{ "f", 0i, {{ "adt", "EvenInt", {{  }} }} }} }}), "i32", true);
        g_3 := l-nth(g_4, 0i);
        g_5 := "i__binop_checked_add"(g_3, 1i, 2147483647i);
        u := [store_value](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "tuple", {{ "i32", "bool" }} }}, g_5);
        g_7 := [load_value](l-nth(m_2, 0i), l+(l-nth(m_2, 1i), {{ {{ "f", 1i, {{ "tuple", {{ "i32", "bool" }} }} }} }}), "bool", false);
        g_6 := l-nth(g_7, 0i);
        u := "i__lang_assert"(not g_6, "overflow check failed");
        goto bb1;
  bb1:  g_9 := [load_value](l-nth(m_2, 0i), l+(l-nth(m_2, 1i), {{ {{ "f", 0i, {{ "tuple", {{ "i32", "bool" }} }} }} }}), "i32", false);
        g_8 := l-nth(g_9, 0i);
        g_11 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, true);
        g_10 := l-nth(g_11, 0i);
        u := [store_value](l-nth(l-nth(g_10, 0i), 0i), l+(l-nth(l-nth(g_10, 0i), 1i), {{ {{ "f", 0i, {{ "adt", "EvenInt", {{  }} }} }} }}), "i32", g_8);
        m_0 := {{}};
        g_12 := m_0;
        ret := g_12;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }});
        u := [free](l-nth(m_2, 0i), l-nth(m_2, 1i), {{ "tuple", {{ "i32", "bool" }} }});
        return
};

spec "EvenInt::test"(pLft_0, self)
[[ "<&mut EvenInt as gilogic::prophecies::Ownable>::own"(#pLft_0, #self, #self_repr) * <lft>(#pLft_0; true) * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj) * (#self == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }}) * (self == #self) * (pLft_0 == #pLft_0) ]]
[["<() as gilogic::prophecies::Ownable>::own"(ret, #ret_repr) * <observation>(true; ) * <lft>(#pLft_0; true)]]
normal

proc "EvenInt::test"(pLft_0, self) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, g_0);
  bb0:  g_2 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, true);
        g_1 := l-nth(g_2, 0i);
        g_4 := [load_value](l-nth(l-nth(g_1, 0i), 0i), l+(l-nth(l-nth(g_1, 0i), 1i), {{ {{ "f", 0i, {{ "adt", "EvenInt", {{  }} }} }} }}), "i32", true);
        g_3 := l-nth(g_4, 0i);
        m_4 := g_3;
        m_5 := ((2i) = (0i));
        g_5 := m_5;
        u := "i__lang_assert"(not g_5, "remainder by zero");
        goto bb1;
  bb1:  m_6 := ((2i) = (-1i));
        m_7 := ((m_4) = (-2147483648i));
        g_6 := m_6;
        g_7 := m_7;
        m_8 := ((g_6) and (g_7));
        g_8 := m_8;
        u := "i__lang_assert"(not g_8, "overflow check failed");
        goto bb2;
  bb2:  g_9 := m_4;
        m_3 := ((g_9) i% (2i));
        g_10 := m_3;
        m_2 := not ((g_10) = (0i));
        g_11 := m_2;
        goto [g_11] bb3 bb4;
  bb3:  g_12 := "EvenInt::test::panic_cold_explicit"();
        m_10 := g_12;
  bb4:  m_0 := {{}};
        g_13 := m_0;
        ret := g_13;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }});
        return
};

proc "EvenInt::test::panic_cold_explicit"() {
  bb0:  g_0 := "core::panicking::panic_explicit"();
        m_0 := g_0;
  rlab: return
};

spec "EvenInt::add_two"(pLft_0, self)
[[ "<&mut EvenInt as gilogic::prophecies::Ownable>::own"(#pLft_0, #self, #self_repr) * <observation>(((l-nth(#self_repr, 0i)) i<= (2147483645i)); ) * <lft>(#pLft_0; true) * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj) * (#self == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }}) * (self == #self) * (pLft_0 == #pLft_0) ]]
[["<() as gilogic::prophecies::Ownable>::own"(ret, #ret_repr) * <observation>(((l-nth(#self_repr, 1i)) = (((l-nth(#self_repr, 0i)) i+ (2i)))); ) * <lft>(#pLft_0; true)]]
normal

proc "EvenInt::add_two"(pLft_0, self) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, g_0);
  bb0:  g_2 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, true);
        g_1 := l-nth(g_2, 0i);
        g_4 := [load_value](l-nth(l-nth(g_1, 0i), 0i), l+(l-nth(l-nth(g_1, 0i), 1i), {{ {{ "f", 0i, {{ "adt", "EvenInt", {{  }} }} }} }}), "i32", true);
        g_3 := l-nth(g_4, 0i);
        m_2 := g_3;
        g_6 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, true);
        g_5 := l-nth(g_6, 0i);
        g_7 := [new_lft]();
        pLft_8 := l-nth(g_7, 0i);
        m_5 := g_5;
        g_8 := m_5;
        g_9 := "EvenInt::add"(pLft_0, g_8);
        m_4 := g_9;
        goto bb1;
  bb1:  g_11 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, true);
        g_10 := l-nth(g_11, 0i);
        g_12 := [new_lft]();
        pLft_9 := l-nth(g_12, 0i);
        m_7 := g_10;
        g_13 := m_7;
        g_14 := "EvenInt::add"(pLft_0, g_13);
        m_6 := g_14;
        goto bb2;
  bb2:  m_3 := {{}};
        g_15 := "gilogic::prophecies::check_obs_sat"();
        m_8 := g_15;
        goto bb3;
  bb3:  g_17 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, true);
        g_16 := l-nth(g_17, 0i);
        g_18 := [new_lft]();
        pLft_10 := l-nth(g_18, 0i);
        m_10 := g_16;
        g_19 := m_10;
        g_20 := "gilogic::prophecies::prophecy_auto_update::<EvenInt>"(pLft_0, g_19);
        m_9 := g_20;
        goto bb4;
  bb4:  g_22 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }}, true);
        g_21 := l-nth(g_22, 0i);
        g_23 := [new_lft]();
        pLft_11 := l-nth(g_23, 0i);
        m_12 := g_21;
        g_24 := m_12;
        g_25 := "gilogic::prophecies::prophecy_resolve::<EvenInt>"(pLft_0, g_24);
        m_11 := g_25;
        goto bb5;
  bb5:  m_0 := {{}};
        g_26 := m_0;
        ret := g_26;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "ref", true, {{ "adt", "EvenInt", {{  }} }} }});
        return
};

